qiskit/tools/visualization/_counts_visualization.py

```diff
@@ -17,35 +17,68 @@
 import matplotlib.pyplot as plt
 
 
-def plot_histogram(data, number_to_keep=False):
+def plot_histogram(data, number_to_keep=False, legend=None, options=None):
     """Plot a histogram of data.
 
-    data is a dictionary of  {'000': 5, '010': 113, ...}
-    number_to_keep is the number of terms to plot and rest is made into a
-    single bar called other values
+    Args:
+        data (list or dict): This is either a list of dictionaries containing:
+            values to represent (ex {'001': 130})
+        number_to_keep (int): DEPRECATED the number of terms to plot and rest
+            is made into a single bar called other values
+        legend(list): A list of strings to use for labels of the data.
+            The number of entries must match the lenght of data (if data is a
+            list or 1 if it's a dict)
+        options (dict): Representation settings containing
+            - number_to_keep (integer): groups max values
+            - show_legend (bool): show legend of graph content
+    Raises:
+        Exception: When legend is provided and the length doesn't match the
+            input data
     """
-    if number_to_keep is not False:
-        data_temp = dict(Counter(data).most_common(number_to_keep))
-        data_temp["rest"] = sum(data.values()) - sum(data_temp.values())
-        data = data_temp
-
-    labels = sorted(data)
-    values = np.array([data[key] for key in labels], dtype=float)
-    pvalues = values / sum(values)
-    numelem = len(values)
-    ind = np.arange(numelem)  # the x locations for the groups
-    width = 0.35  # the width of the bars
+    if options is None:
+        options = {}
+
+    if isinstance(data, dict):
+        data = [data]
+        if legend and len(legend) != 1:
+            raise Exception("Length of legendL %s doesn't match number of "
+                            "input executions: %s" % (len(legend), 1))
+    else:
+        if legend and len(legend) != len(data):
+            raise Exception("Length of legendL %s doesn't match number of "
+                            "input executions: %s" % (len(legend), len(data)))
+
     _, ax = plt.subplots()
-    rects = ax.bar(ind, pvalues, width, color='seagreen')
-    # add some text for labels, title, and axes ticks
-    ax.set_ylabel('Probabilities', fontsize=12)
-    ax.set_xticks(ind)
-    ax.set_xticklabels(labels, fontsize=12, rotation=70)
-    ax.set_ylim([0., min([1.2, max([1.2 * val for val in pvalues])])])
-    # attach some text labels
-    for rect in rects:
-        height = rect.get_height()
-        ax.text(rect.get_x() + rect.get_width() / 2., 1.05 * height,
-                '%f' % float(height),
-                ha='center', va='bottom')
+    for item, execution in enumerate(data):
```

Comment:
Non-blocking: do you think is it feasible or even useful to allow customization of graph `width`, graph `height` and `sort` to equate the options dictionary passed to this function to that supported by the `iplot_function`?

Reply:
It should be, and I thought about that when I was refactoring things. But I didn't want to change things too much here. After this merges I'll work on updating this to add the missing options.